openmoc/materialize.py

```diff
@@ -355,3 +360,328 @@ def load_openmc_mgxs_lib(mgxs_lib, geometry=None):
+    for i in range(max_domain_iters):
+        for domain_id in sph_domains:
+            py_printf('NORMAL', 'SPH outer iteration %d: %s %d',
+                                i, mgxs_lib.domain_type.capitalize(), domain_id)
+
+            inner_sph = np.ones((num_domains, num_groups))
+
+            for j in range(max_fix_src_iters):
+
+                # Run fixed source calculation with suppressed output
+                if throttle_output and log_level <= openmoc.WARNING:
+                    openmoc.set_log_level('WARNING')
+
+                # Fixed source calculation
+                solver.computeFlux()
+
+                # Restore log output level
+                if throttle_output and log_level <= openmoc.NORMAL: 
+                    openmoc.set_log_level('NORMAL')
+
+                # Extract the FSR scalar fluxes
+                fsr_fluxes = get_scalar_fluxes(solver)
+
+                # Compute the domain-averaged flux in each energy group
+                for k, domain in enumerate(mgxs_lib.domains):
+                    domain_fluxes = fsr_fluxes[fsrs_to_domains == domain.id, :]
+                    openmoc_fluxes[k, :] = np.mean(domain_fluxes, axis=0)
+
+                    # Store index for the domain in the array of SPH factors
+                    if domain.id == domain_id:
+                        sph_index = k
+
+                # Compute SPH factors
+                old_inner_sph = np.copy(inner_sph)
+                inner_sph = openmc_fluxes / openmoc_fluxes
+                inner_sph = np.nan_to_num(inner_sph)
+                inner_sph[inner_sph == 0.0] = 1.0
+
+                # Compute SPH factor residuals
+                inner_res = np.abs((inner_sph - old_inner_sph) / old_inner_sph)
+                inner_res = np.nan_to_num(inner_res)
+
+                # Report maximum SPH factor residual
+                py_printf('NORMAL', 'SPH inner iteration %d:\tres = '
+                                    '%1.3e', j, inner_res.max())
+
+                # Update multi-group cross sections with SPH factors
+                sph_mgxs_lib = _apply_sph_factors(mgxs_lib, inner_sph[sph_index],
+                                                  sph_domains[domain_id])
+
+                # Load the new MGXS library data into the OpenMOC geometry
+                load_openmc_mgxs_lib(sph_mgxs_lib, geometry)
+
+                # Check max SPH factor residual for this domain for convergence
+                if inner_res[sph_index].max() < sph_tol:
+                    break
+
+            mgxs_lib = sph_mgxs_lib
+
+        # Check max SPH factor residuals for domains of interest for convergence
+        if inner_res[sph_indices].max() < sph_tol:
+            break
+
+    # Warn user if SPH factors did not converge
+    else:
+        py_printf('WARNING', 'SPH factors did not converge')
+
+    # Compute the final SPH factors for all domains and energy groups
+    final_sph = openmc_fluxes / openmoc_fluxes
```

Comment:
You raise a very good point here - one which I considered when designing the SPH scheme. I agree that it wouldn't at all far-fetched to imagine subgroup and/or ultrafine capabilities which also make use of an SPH scheme. However, I settled upon the current design for the following three reasons:

* It is now so easy (knock on wood) to generate MGXS in OpenMC and plug them into OpenMC that I see it as becoming the de facto standard approach to simulation workflows with OpenMOC for the foreseeable future.

* There is no fundamental software issue precluding subgroup / ultrafine scheme implementations in OpenMOC. The fundamental issue is a people one - no one has stepped up to the plate yet (no offense intended), and I don't see anyone in the pipeline who will implement a "production" version in the future. *If* someone does need these capabilities, then we can discuss how best to better abstract the SPH calculator at that time.

* Related to my last point, it is very difficult to design code abstractions for an ambiguous or moving target. I think this point is related to Donald Knuths' truism that "premature optimization is the root of all evil." I know the OpenMC MGXS generation module very well, and can hence design an SPH customized for it, but don't have time to expend trying to design code for some unknown and non-existent type of MGXS library.

* Lastly, any future developer(s) of subgroup / ultrafine features will need to design scalable data structure(s) to store and retrieve MGXS (and scalar fluxes!) for arbitrary group structures on arbitrary unstructured meshes, yada, yada, yada. My guess is that such a data structure may look a lot like my `openmc.mgxs.MGXS` and `openmc.mgxs.Library` classes. As a result, I'll cross my fingers now and just say that the SPH factor calculator *should* be easily extensible to alternative MGXS library types if an when someone comes along with some time to burn to make it work.

That's my long-winded way of saying that I'm not going to bite on this idea at this point :-) But if you have any thoughts about how this might be better abstracted, write them down here now as a historical record. Who knows, they might be useful in a few years.